anitya/lib/backends/github.py

```diff
@@ -59,18 +65,168 @@ def get_versions(cls, project):
             if url.endswith('/'):
-                url = project.homepage[:-1]
-            url += '/tags'
+                url = url[:-1]
         else:
             raise AnityaPluginException(
                 'Project %s was incorrectly set-up' % project.name)
 
-        return get_versions_by_regex(url, REGEX, project)
+        if url:
+            (owner, repo) = url.split('/')
+
+        if (not owner) or (not repo):
+            raise AnityaPluginException(
+                'Project %s was incorrectly set-up. \
+                Can\'t parse owner and repo' % project.name)
+
+        query = cls.prepare_query(owner, repo)
+
+        try:
+            resp = cls.call_url_post(API_URL, query, use_token=True)
+        except Exception as err:
+            _log.debug('%s ERROR: %s' % (project.name, str(err)))
+            raise AnityaPluginException(
+                'Could not call : "%s" of "%s", with error: %s' % (
+                    url, project.name, str(err)))
+
+        if resp.ok:
+            json = resp.json()
+        else:
+            raise AnityaPluginException(
+                '%s: Server responded with status "%s": "%s"' % (
+                    project.name, resp.status_code, resp.reason))
+
+        if 'errors' in json:
+            error_str = ''
+            for error in json['errors']:
+                error_str += '"%s": "%s"\n' % (
+                    error['type'], error['message'])
+            raise AnityaPluginException(
+                '%s: Server responded with following errors\n%s' % (
+                    project.name, error_str))
+
+        total_count = json['data']['repository']['refs']['totalCount']
+
+        _log.debug('Received %s tags for %s' % (total_count, project.name))
+
+        # TODO: Save cursor to project and use it to fetch only newer versions
+        # cursor = json['data']['repository']['refs']['edges']['cursor']
+
+        remaining = json['data']['rateLimit']['remaining']
+        reset_time = json['data']['rateLimit']['resetAt']
+        _log.debug('Github API ratelimit remains %s, will reset at %s UTC' % (
+                remaining, reset_time))
+
+        versions = []
+
+        for edge in json['data']['repository']['refs']['edges']:
+            version = edge['node']['name']
+            if project.version_prefix is not None and \
+                    version.startswith(project.version_prefix):
+                version = version[len(project.version_prefix):]
+            elif version.startswith('v'):
+                version = version[1:]
+            # TODO: Save commit url to DB and make versions clickable on frontend
+            # commit_url = edge['node']['target']['commitUrl']
+            versions.append(version)
+
+        return versions
+
+    @classmethod
+    def prepare_query(cls, owner, repo, after=''):
+        ''' Method for preparing GraphQL query for specified repository
+
+        :arg owner: owner of the repository
```

Comment:
I find the ["Google"](http://www.sphinx-doc.org/en/1.7/ext/example_google.html) style of comments more pleasant to write, you might as well.

Reply:
I used the style, that was there before, but I will change it.